src/desul/array.hpp

```diff
@@ -0,0 +1,301 @@
+#if !defined(DESUL_ARRAY_HPP)
+#define DESUL_ARRAY_HPP
+
+#include "desul/Macros.h"
+#include "desul/utility.hpp"
+
+#include <cstddef>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+namespace desul {
+   ///
+   /// Provides a portable std::array-like class.
+   /// 
+   /// Differences from std::array are listed below:
+   /// - No __device__ "at" method (exceptions are not GPU friendly)
+   /// - Calling front or back is a compile error when size() == 0
+   ///   (instead of undefined behavior at run time)
+   /// - No reverse iterators have been implemented yet
+   ///
+   template <class T, std::size_t N>
+   struct array {
+      using value_type = T;
+      using size_type = std::size_t;
+      using difference_type = std::ptrdiff_t;
+      using reference = value_type&;
+      using const_reference = const value_type&;
+      using pointer = value_type*;
+      using const_pointer = const value_type*;
+      using iterator = pointer;
+      using const_iterator = const_pointer;
+      using reverse_iterator = pointer;
+      using const_reverse_iterator = const_pointer;
+
+      constexpr reference at(size_type pos) {
+         if (pos >= size()) {
+            throw std::out_of_range{"desul::array::at detected out of range access"};
+         }
```

Comment:
Should we think about using assert in device code instead of throwing as a away to check bounds?